### PR TITLE
fix(zone): suggestions for #7891 zone makeOnce etc

### PR DESCRIPTION
**Pull request by erights**
Suggestions for #7891 

My review comments on each change explains it.

### CODE REVIEW

File: packages/internal/src/is-passable.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { passStyleOf } from '@endo/pass-style';
+
+/**
+ * Is `specimen` Passable? This returns true iff `passStyleOf(specimen)`
+ * returns a string. This returns `false` iff `passStyleOf(specimen)` throws.
+ * Under no normal circumstance should `isPassable(specimen)` throw.
+ *
+ * TODO implement an isPassable that does not rely on try/catch, and
+ * move it to @endo/pass-style.
+ * This implementation is just a standin until then
```
Comment (erights) [inline packages/internal/src/is-passable.js]: I thought that `@agoric/internal` was a decent temporary place for this until it is moved to `@endo/pass-style`. But it turns out that it creates a dependency cycle (thanks CI) because `@agoric/internal` currently depends on `@agoric/zone`. 

IIUC, this temporarily violates the intention that `@agoric/internal` is at the bottom of the agoric-sdk dependency graph. But seeing as we expect most of `@agoric/zone` to migrate to `@endo/zone`, perhaps we should not fix `@agoric/internal` to stop depending on Zones?

In any case, for #7891 , I suggest that this temporary kludgy `isPassable` be moved into `@agoric/zone` itself, while keeping the TODO note above.

File: packages/zone/src/durable.js [hunk 2]
```diff
@@ -27,13 +26,7 @@ const { Fail } = assert;
...
  * @returns {boolean}
  */
-const isStorable = specimen => {
-  try {
-    return canBeDurable(specimen);
-  } catch (_e) {
-    return false;
-  }
-};
+const isStorable = specimen => isPassable(specimen) && canBeDurable(specimen);
```
Comment (erights) [inline packages/zone/src/durable.js:29]: It *should* be the case that `canBeDurable` throws only if `specimen` is not Passable. If so, this is a safe simplification. But it does indeed make `isStorable`'s no-throw guarantee more brittle, so feel free to restore the try-catch. But even then, I think you should keep the conjunction.

File: packages/zone/src/durable.js [hunk 3]
```diff
@@ -54,7 +47,7 @@ const attachDurableStores = getBaggage => {
     provideDurableWeakMapStore(getBaggage(), label, options);
 
   /** @type {import('.').Stores} */
-  return Far('durableStores', {
+  return harden({
```
Comment (erights) [inline packages/zone/src/durable.js]: I am not actually suggesting any of the changes of stores and zones from `Far` to `harden`. Rather, I did it in this PR just to make them non-passable, so I could test that I fixed the problems of where they sometimes try to store themselves.

At https://github.com/Agoric/agoric-sdk/pull/7972 I took steps to make stores and zones into self-storing exos. This failed for reasons I'll explain there. Given that failure, I think that we should still treat stores and zones as non-self storing.

So for all these changes of `Far` to `harden`: change not suggested.

File: packages/zone/src/durable.js [hunk 4]
```diff
@@ -103,7 +96,7 @@ export const makeDurableZone = (baggage, baseLabel = 'durableZone') => {
     return makeDurableZone(subBaggage, `${baseLabel}.${label}`);
```
Comment (erights) [inline packages/zone/src/durable.js]: I like what durable zone is doing for `subZone` --- making the backing store be the stores child of the parent, and leaving the wrapping zone unstored. Amusingly, were `subZone` intended not to be collision protected, this would be a problem in that every time we call `subZone` with the same parent and label, we'd get a new subZone object identity. But since we can only do so once per incarnation anyway, this case cannot arise.

So no change suggested here. I'm commenting on it because part of the point of this PR is to more closely follow the precedent set by durable zones.

File: packages/zone/src/heap.js [hunk 5]
```diff
@@ -1,28 +1,25 @@
...
 import { agoricVatDataKeys as keys } from './keys.js';
 
 /**
  * @type {import('.').Stores}
  */
-const detachedHeapStores = Far('heapStores', {
+const detachedHeapStores = harden({
   detached: () => detachedHeapStores,
-  isStorable: _specimen => true,
+  isStorable: isPassable,
```
Comment (erights) [inline packages/zone/src/heap.js:22]: In an unpublished draft I tried reproducing in heap.js the `isStorable` check that #7891 has for virtual isStorable, but using a heap store rather than a virtual store. As I expected, the heap store refuses to store non-passables. Thus, I think the heap zone should consider non-passables to not be heap storable. 

As explained at https://github.com/endojs/endo/issues/1648 , heap exos will currently store non-passables. But I filed the issue because I consider that a bug. Your PR should proceed on the heap store precedent, assuming this bug is eventually fixed.

File: packages/zone/src/heap.js [hunk 6]
```diff
@@ -38,13 +35,15 @@ const detachedHeapStores = Far('heapStores', {
  */
 export const makeHeapZone = (baseLabel = 'heapZone') => {
   const { makeOnce, wrapProvider } = makeOnceKit(baseLabel, detachedHeapStores);
-  return Far('heapZone', {
+
+  const subZoneProvider = (label, _options) =>
```
Comment (erights) [inline packages/zone/src/heap.js]: Please rename to be consistent with whatever naming convention you have for these encapsulated provider functions. I am not suggesting any deviation from those conventions.

File: packages/zone/src/heap.js [hunk 7]
```diff
@@ -38,13 +35,15 @@ const detachedHeapStores = Far('heapStores', {
...
+    makeHeapZone(`${baseLabel}.${label}`);
+
+  return harden({
     exo: wrapProvider(makeExo, keys.exo),
```
Comment (erights) [inline packages/zone/src/heap.js:43]: This works because `wrapProvider` does not enforce that the provided thing is storable, whereas `makeOnce` necessarily does.

File: packages/zone/src/heap.js [hunk 8]
```diff
@@ -38,13 +35,15 @@ const detachedHeapStores = Far('heapStores', {
...
     exoClass: wrapProvider(defineExoClass, keys.exoClass),
     exoClassKit: wrapProvider(defineExoClassKit, keys.exoClassKit),
-    subZone: (label, _options) => {
-      return makeOnce(label, () => makeHeapZone(`${baseLabel}.${label}`));
-    },
+    subZone: wrapProvider(subZoneProvider),
```
Comment (erights) [inline packages/zone/src/heap.js]: This works because `wrapProvider` does not enforce that the provided thing is storable, whereas `makeOnce` necessarily does.

File: packages/zone/src/virtual.js [hunk 9]
```diff
@@ -25,7 +24,7 @@ const initEmpty = harden(() => emptyRecord);
  *
  * @type {import('.').Zone['exo']}
  */
-const defineVirtualExo = (
+const makeVirtualExo = (
```
Comment (erights) [inline packages/zone/src/virtual.js:27]: The `make` vs `define` convention was introduced to distinguish functions that make things vs functions that define and return make functions. This name change is more consistent with the matrix this is completing.

File: packages/zone/src/virtual.js [hunk 10]
```diff
@@ -45,43 +44,10 @@ const defineVirtualExo = (
...
-  }
-};
-harden(isStorable);
-
 /** @type {import('.').Stores} */
-export const detachedVirtualStores = Far('virtualStores', {
+export const detachedVirtualStores = harden({
   detached: () => detachedVirtualStores,
-  isStorable,
+  isStorable: isPassable,
```
Comment (erights) [inline packages/zone/src/virtual.js:50]: I'm suggesting this change because I believe that anything `Passable` can be stored as a value in a virtual store, so an `isPassable` check will always be equivalent to the virtual `isStorable` in the #7891 code in the left panel above.

@FUDCo @warner Is this true?

File: packages/zone/test/test-make-once.js [hunk 11]
```diff
@@ -30,35 +30,29 @@ test.before(t => {
 /**
  * @param {import('ava').Assertions} t
  * @param {Zone} rootZone
- * @param {boolean} [heapOnlyZone]
  */
-const testOnce = (t, rootZone, heapOnlyZone = false) => {
+const testOnce = (t, rootZone) => {
```
Comment (erights) [inline packages/zone/test/test-make-once.js:34]: I take it as a positive symptom that in this PR the test cases be more regular between the heap vs virtual and durable.

Comment (erights) [reply packages/zone/test/test-make-once.js:34]: I like the idea of a `testOnce` for testing behavior that should be in common across zones. But these tests are a bit thin.

File: packages/internal/src/is-passable.js (hunk 1 above)
Comment (erights) [reply packages/internal/src/is-passable.js]: @turadg just so you're aware of the dependency violation. Note that I suggest above why IMO it does not need to be fixed.

**Review by michaelfig — COMMENTED** [review]
Thanks for the thorough examination of #7891.

I've had a look at these changes and explanations, and they seem well-founded to me.

I saw that you had a few open questions, and that you suggested moving `is-passable.js` to the zone package.  Would you please ping me again when you're satisfied with this PR's state?  Then I would be happy to take another look and merge.

File: packages/zone/src/durable.js [hunk 12]
```diff
@@ -59,7 +59,7 @@ const attachDurableStores = getBaggage => {
 };
 
 /** @type {import('.').Stores} */
-export const detachedDurableStores = attachDurableStores(() =>
+const detachedDurableStores = attachDurableStores(() =>
```
Comment (erights) [inline packages/zone/src/durable.js:62]: Removed the `export` because `detachedHeapStores` is not exported, and I do not see any reason it should be.

File: packages/zone/src/durable.js [hunk 13]
```diff
@@ -113,5 +113,3 @@ export const makeDurableZone = (baggage, baseLabel = 'durableZone') => {
   });
 };
 harden(makeDurableZone);
-
-export { M };
```
Comment (erights) [inline packages/zone/src/durable.js:124]: Why was this here? (And in the other two as well)

File: packages/zone/src/virtual.js [hunk 14]
```diff
@@ -45,7 +45,7 @@ const makeVirtualExo = (
 };
 
 /** @type {import('.').Stores} */
-export const detachedVirtualStores = harden({
+const detachedVirtualStores = Far('virtualStores', {
```
Comment (erights) [inline packages/zone/src/virtual.js:48]: Removed the `export` because `detachedHeapStores` is not exported, and I do not see any reason it should be.

File: packages/zone/src/virtual.js [hunk 15]
```diff
@@ -100,13 +66,14 @@ export const makeVirtualZone = (baseLabel = 'virtualZone') => {
     baseLabel,
     detachedVirtualStores,
   );
-  return Far('VirtualZone', {
-    exo: wrapProvider(defineVirtualExo, keys.exo),
+  const subZoneProvider = (label, _options) =>
+    makeVirtualZone(`${baseLabel}.${label}`);
+
+  return Far('virtualZone', {
```
Comment (erights) [inline packages/zone/src/virtual.js:77]: Changed the case, because all the other ones started with lower case.

File: packages/zone/src/durable.js (hunk 13 above)
Comment (erights) [reply packages/zone/src/durable.js:124]: Well, I found out why --- `M` is imported from `@agoric/zone`. Rather than restore this, which seems very weird to me, instead I fixed the importers to import `M` from `@endo/patterns` instead.

If you don't like this, feel free to instead restore the exporting of `M` from `@agoric/zone`. Though, perhaps better to export it from the exporting files (index.js, etc) rather than exporting it from the substantive source code files.

**Comment by erights** [comment]
> Would you please ping me again when you're satisfied with this PR's state? Then I would be happy to take another look and merge.

Ping ;)

File: packages/zone/src/heap.js (hunk 8 above)
Comment (michaelfig) [reply packages/zone/src/heap.js]: Ah!  Okay, that makes sense.

File: packages/zone/src/durable.js (hunk 2 above)
Comment (michaelfig) [reply packages/zone/src/durable.js:29]: I think it's okay to be a little bit brittle, since that will avoid masking problems in changes to the `canBeDurable` implementation.

File: packages/zone/src/virtual.js (hunk 10 above)
Comment (michaelfig) [reply packages/zone/src/virtual.js:50]: @mhofman says:
> I think promises are currently an exception... or at least bad things happen if we use promises in virtual stores

File: packages/zone/test/test-make-once.js (hunk 11 above)
Comment (michaelfig) [reply packages/zone/test/test-make-once.js:34]: > these tests are a bit thin.

Yeah, `testOnce` is really only meant to test the `makeOnce` behaviour.  I still need to make more comprehensive tests (elsewhere) that explore more of the API.

File: packages/zone/src/heap.js [hunk 16]
```diff
@@ -38,13 +35,15 @@ const detachedHeapStores = Far('heapStores', {
  */
 export const makeHeapZone = (baseLabel = 'heapZone') => {
   const { makeOnce, wrapProvider } = makeOnceKit(baseLabel, detachedHeapStores);
+
+  const subZoneProvider = (label, _options) =>
```
Comment (michaelfig) [inline packages/zone/src/heap.js]:
```suggestion
  /**
   * @param {string} label
   * @param {any} _options
   */
  const provideSubZone = (label, _options) =>
```

File: packages/zone/src/heap.js [hunk 17]
```diff
@@ -38,13 +35,15 @@ const detachedHeapStores = Far('heapStores', {
...
+    makeHeapZone(`${baseLabel}.${label}`);
+
   return Far('heapZone', {
     exo: wrapProvider(makeExo, keys.exo),
     exoClass: wrapProvider(defineExoClass, keys.exoClass),
     exoClassKit: wrapProvider(defineExoClassKit, keys.exoClassKit),
-    subZone: (label, _options) => {
-      return makeOnce(label, () => makeHeapZone(`${baseLabel}.${label}`));
-    },
+    subZone: wrapProvider(subZoneProvider),
```
Comment (michaelfig) [inline packages/zone/src/heap.js]:
```suggestion
    subZone: wrapProvider(provideSubZone),
```

File: packages/zone/src/virtual.js [hunk 18]
```diff
@@ -100,13 +66,14 @@ export const makeVirtualZone = (baseLabel = 'virtualZone') => {
     baseLabel,
     detachedVirtualStores,
   );
-  return Far('VirtualZone', {
-    exo: wrapProvider(defineVirtualExo, keys.exo),
+  const subZoneProvider = (label, _options) =>
```
Comment (michaelfig) [inline packages/zone/src/virtual.js]:
```suggestion

  /**
   * @param {string} label
   * @param {any} _options
   */
  const provideSubZone = (label, _options) =>
```

File: packages/zone/src/virtual.js [hunk 19]
```diff
@@ -100,13 +66,14 @@ export const makeVirtualZone = (baseLabel = 'virtualZone') => {
...
+    makeVirtualZone(`${baseLabel}.${label}`);
+
+  return Far('virtualZone', {
+    exo: wrapProvider(makeVirtualExo, keys.exo),
     exoClass: wrapProvider(defineVirtualExoClass, keys.exoClass),
     exoClassKit: wrapProvider(defineVirtualExoClassKit, keys.exoClassKit),
-    subZone: (label, _options) => {
-      return makeOnce(label, () => makeVirtualZone(`${baseLabel}.${label}`));
-    },
+    subZone: wrapProvider(subZoneProvider),
```
Comment (michaelfig) [inline packages/zone/src/virtual.js]:
```suggestion
    subZone: wrapProvider(makeSubZone),
```

**Review by michaelfig — COMMENTED** [review]
Some of the suggested `@param` type annotations should get CI to pass.  Please accept my other code suggestions, and then I think the only remaining question is how to exclude promises from `isStorable`.

File: packages/zone/src/virtual.js (hunk 10 above)
Comment (erights) [reply packages/zone/src/virtual.js:50]: My hypothesis is that the same thing happens as for `Far` objects --- they get "exported" and an "import" of this internal export gets stored into virtual storage.

Someone should try ;)

Attn @FUDCo @warner @mhofman

File: packages/zone/src/heap.js (hunk 17 above)
Comment (erights) [reply packages/zone/src/heap.js]: This renaming does not sound good to me because the argument function itself is not a `provide*` function according to our convention. The function returned by `wrapProvider` is a `provide*` function according to our convention.

File: packages/zone/src/heap.js (hunk 16 above)
Comment (erights) [reply packages/zone/src/heap.js:43]: Done

Comment (erights) [reply packages/zone/src/heap.js]: Param types added. Renaming not for stated reasons.

File: packages/zone/src/virtual.js (hunk 18 above)
Comment (erights) [reply packages/zone/src/virtual.js]: Param types added. Renaming not for stated reasons.

File: packages/zone/src/heap.js (hunk 17 above)
Comment (michaelfig) [reply packages/zone/src/heap.js]: Then maybe `makeSubZone` instead?

File: packages/internal/src/is-passable.js (hunk 1 above)
Comment (turadg) [reply packages/internal/src/is-passable.js]: Your understanding is correct and I endorse your conclusion. :)

File: packages/zone/src/heap.js (hunk 17 above)
Comment (erights) [reply packages/zone/src/heap.js]: Done

File: packages/zone/src/virtual.js (hunk 10 above)
Comment (michaelfig) [reply packages/zone/src/virtual.js:50]: > Someone should try ;)

I'll gladly write a test for this in #7891 after the current PR lands.

**Review by michaelfig — APPROVED** [review]
LGTM!  Thanks for clearing up these unclear bits.
